contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,728 @@
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector
+
+	// Describes the rules on how changes are applied to the pods.
+	// +optional
+	UpdatePolicy PodUpdatePolicy
+
+	// Controls how the autoscaler computes recommended resources.
+	// +optional
+	ResourcePolicy PodResourcePolicy
+}
+
+// VerticalPodAutoscalerStatus describes the runtime state of the autoscaler.
+type VerticalPodAutoscalerStatus {
+	// The time when the status was last refreshed.
+	LastUpdateTime metav1.Time
+	// The most recently computed amount of resources recommended by the
+	// autoscaler for the controlled pods.
+	// +optional
+	Recommendation RecommendedPodResources	
+	// A free-form human readable message describing the status of the autoscaler.
+	StatusMessage string
+}
+```
+
+The complete API definition is included [below](#complete_vpa_api_object_definition).
+
+#### Label Selector ####
+The label selector determines which Pods will be scaled according to the given
+VPA policy. The Recommender will aggregate signals for all Pods matched by a
+given VPA, so it is important that the user set labels to group similarly
+behaving Pods under one VPA.
+
+It is yet to be determined how to resolve conflicts, i.e. when the Pod is
+matched by more than one VPA (this is not a VPA-specific problem though).
+
+#### Update Policy ####
+The update policy controls how VPA applies changes. In MVP it consists of a
+single field `mode` that enables the feature.
+
+```json
+"updatePolicy" {
+  "mode": "",
+}
+```
+
+Mode can be set to one of the following:
+
+1. `"Initial"`: VPA only assigns resources on Pod creation and does not
+   change them during lifetime of the Pod.
+2. `"Auto"` (default): VPA assigns resources on Pod creation and
+   additionally can update them during lifetime of the Pod, including evicting /
+   rescheduling the Pod.
+3. `"Off"`: VPA never changes Pod resources. The recommender still sets the
+   recommended resources in the VPA object. This can be used for a “dry run”.
+
+To disable VPA updates the user can do any of the following: (1) change the
+updatePolicy to `"Off"` or (2) delete the VPA or (3) change the Pod labels to no
+longer match the VPA selector.
+
+Note: disabling VPA prevents it from doing further changes, but does not revert
+resources of the running Pods, until they are updated.
+For example, when running a Deployment, the user would need to perform an update
+to revert Pod to originally specified resources.
+
+#### Resource Policy ####
+The resources policy controls how VPA computes the recommended resources.
+In MVP it consists of (optional) lower and upper bound on the request of each
+container.
+The resources policy could later be extended with additional knobs to let the
+user tune the recommendation algorithm to their specific use-case.
+
+#### Recommendation ####
+The VPA resource has an output-only field keeping a recent recommendation,
```

Comment:
What are the scale implications of writing this (i.e. if we have 1k VPA, how often do we write this?)

Reply:
It may be written every 5-10 minutes (if the recommended resources change), which I suppose shouldn't be an issue?
The freshness of this data is not so critical: both the Admission Controller and the Updater always obtain the most recent recommendation directly from the Recommender API, before updating the pod. Only if the Recommender is down, they may use the recommendation cached in the VPA object.
The only concern is fast reaction to OOMs. The updater should be able to detect OOM-ing containers and call the Recommender API more frequently to ensure fast reaction time.